src/bvals/cc/bvals_cc_in_one.cpp

```diff
@@ -0,0 +1,553 @@
+      // Need to update the buffer sizes so that it matches the packed Variables.
+      assert(rc->GetCellVariableVector().size() == 1);
+      auto *bd_var_ = rc->GetCellVariableVector()[0]->vbvar->GetPBdVar();
+      if (bd_var_->sflag[nb.bufid] == parthenon::BoundaryStatus::completed) continue;
+      buffers_used += 1;
+    }
+  }
+  Kokkos::Profiling::popRegion(); // Reset boundaries
+
+  auto boundary_info = md->GetSendBuffers();
+
+  if (!boundary_info.is_allocated() || true) {
+    Kokkos::Profiling::pushRegion("Create bndinfo array");
+
+    boundary_info = BufferCache_t("send_boundary_info", buffers_used);
+    auto boundary_info_h = Kokkos::create_mirror_view(boundary_info);
+
+    // now fill the buffer information
+    int b = 0; // buffer index
+    for (auto block = 0; block < md->NumBlocks(); block++) {
+      auto &rc = md->GetBlockData(block);
+      auto pmb = rc->GetBlockPointer();
+
+      int mylevel = pmb->loc.level;
+      for (int n = 0; n < pmb->pbval->nneighbor; n++) {
+        parthenon::NeighborBlock &nb = pmb->pbval->neighbor[n];
+        // TODO(?) currently this only works for a single "Variable" per container.
+        // Need to update the buffer sizes so that it matches the packed Variables.
+        assert(rc->GetCellVariableVector().size() == 1);
+        auto *bd_var_ = rc->GetCellVariableVector()[0]->vbvar->GetPBdVar();
+        if (bd_var_->sflag[nb.bufid] == parthenon::BoundaryStatus::completed) continue;
+
+        auto &si = boundary_info_h(b).si;
+        auto &ei = boundary_info_h(b).ei;
+        auto &sj = boundary_info_h(b).sj;
+        auto &ej = boundary_info_h(b).ej;
+        auto &sk = boundary_info_h(b).sk;
+        auto &ek = boundary_info_h(b).ek;
+
+        IndexDomain interior = IndexDomain::interior;
+        auto &var_cc = rc->GetCellVariableVector()[0]->data;
+        if (nb.snb.level == mylevel) {
+          const parthenon::IndexShape &cellbounds = pmb->cellbounds;
+          CalcIndicesLoadSame(nb.ni.ox1, si, ei, cellbounds.GetBoundsI(interior));
+          CalcIndicesLoadSame(nb.ni.ox2, sj, ej, cellbounds.GetBoundsJ(interior));
+          CalcIndicesLoadSame(nb.ni.ox3, sk, ek, cellbounds.GetBoundsK(interior));
+          boundary_info_h(b).var = var_cc.Get<4>();
+
+        } else if (nb.snb.level < mylevel) {
+          const IndexShape &c_cellbounds = pmb->c_cellbounds;
+          // "Same" logic is the same for loading to a coarse buffer, just using
+          // c_cellbounds
+          CalcIndicesLoadSame(nb.ni.ox1, si, ei, c_cellbounds.GetBoundsI(interior));
+          CalcIndicesLoadSame(nb.ni.ox2, sj, ej, c_cellbounds.GetBoundsJ(interior));
+          CalcIndicesLoadSame(nb.ni.ox3, sk, ek, c_cellbounds.GetBoundsK(interior));
+
+          auto &coarse_buf = rc->GetCellVariableVector()[0]->vbvar->coarse_buf;
+          pmb->pmr->RestrictCellCenteredValues(var_cc, coarse_buf, 0, Nv - 1, si, ei, sj,
+                                               ej, sk, ek);
```

Comment:
If I recall, this is done variable-by-variable so this function needs to be re-written to use a pack.

Reply:
Correct, though I think we should first do some profiling to determine how much of a performance impact this is.
Given that these are way fewer calls (only a subset of the meshblocks will call this on a subset of the blocks) we may get away with relying on executing the small pack kernels in parallel on the device with a negligible impact of the launch overhead.